Split the PubSub sample down into smaller samples

TwistedTwigleg opened the pull request:
*Description of changes:*

Splits the pub-sub sample down into multiple smaller samples that connect using a single method, rather than having one large pub-sub sample that can connect with multiple methods. This makes it easier for users to see how to connect using each method. Comments were added to show when the pub-sub part of the samples begin, to make it easier to separate the connection part from the pub-sub part.

Also adds helper functions to the CommandLineUtils function for making a MQTT connection without needing to manually set it up, removing the MQTT connection code duplication across samples and instead allowing them to focus on showing the key feature of the sample.

Finally, this PR also refactors the samples a little to make them a bit more compact and adds a missing parameter in the Greengrass Basic Discovery sample that I mistakenly removed in the command line module PR.

_________

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.*

graebm commented on samples/identity/fleet_provisioning/main.cpp:
```diff
@@ -85,76 +76,17 @@ int main(int argc, char *argv[])
-     * and its underlying memory is managed by the client.
-     */
-    auto connection = mqttClient.NewConnection(clientConfig);
+    apiHandle.InitializeLogging(Aws::Crt::LogLevel::Trace, stderr);
```
`Trace` logging is insanely verbose, maybe just do `Error` logging by default

graebm commented on samples/README.md:
```diff
@@ -1,6 +1,7 @@
 # Sample apps for the AWS IoT Device SDK for C++ v2
 
 * [Basic MQTT Pub-Sub](#basic-mqtt-pub-sub)
+* [Websocket MQTT Pub-Sub](#websocket-mqtt-pub-sub)
```
I vote that we don't even bother with publish/subscribe in the connection samples, just do: connect, disconnect, exit.

There's quite a bit of code devoted to subscribing, sending and receiving an exact number of messages, letting users customize the path & message, etc

We can have make the pub-sub sample its own thing like the jobs and shadow sample and just have it use the default connection setup

graebm commented on samples/mqtt/websocket_pub_sub/main.cpp:
```diff
@@ -0,0 +1,302 @@
+    /*********************** Parse Arguments ***************************/
+    Utils::CommandLineUtils cmdUtils = Utils::CommandLineUtils();
+    cmdUtils.RegisterProgramName("basic_pub_sub");
+    cmdUtils.AddCommonMQTTCommands();
+    cmdUtils.RemoveCommand("cert");
+    cmdUtils.RemoveCommand("key");
```
"cert" and "key" aren't actually all that common, they're just for direct MQTT connections using mutual TLS

graebm commented on samples/greengrass/basic_discovery/main.cpp:
```diff
@@ -53,20 +43,15 @@ int main(int argc, char *argv[])
+    String thingName = cmdUtils.GetCommandRequired("thing_name");
+    String caFile = cmdUtils.GetCommandOrDefault("ca_file", "");
+    String keyPath = cmdUtils.GetCommandOrDefault("key", "");
+    String region = cmdUtils.GetCommandOrDefault("region", "us-east-1");
```
remove the "us-east-1" default
AWS has been trying to move away from having this as a hardcoded default

graebm commented on samples/mqtt/x509_pub_sub/main.cpp:
```diff
@@ -0,0 +1,366 @@
+    cmdUtils.AddCommonX509Commands();
+    cmdUtils.RemoveCommand("x509");
+    cmdUtils.AddCommonWebsocketCommands();
+    cmdUtils.RemoveCommand("use_websocket");
```
why does the `use_websocket` option even exist? if the samples are for specific connection types?

graebm commented on samples/mqtt/x509_pub_sub/main.cpp:
```diff
@@ -0,0 +1,366 @@
+    cmdUtils.AddCommonProxyCommands();
+    cmdUtils.AddCommonTopicMessageCommands();
+    cmdUtils.AddCommonX509Commands();
+    cmdUtils.RemoveCommand("x509");
```
why does the `x509` option even exist, if the samples are for specific connection types

TwistedTwigleg commented:
Thanks for the review! I adjusted the code accordingly 👍

graebm commented on samples/mqtt/pkcs11_connect/main.cpp:
```diff
@@ -199,99 +179,24 @@ int main(int argc, char *argv[])
-        exit(-1);
-    }
+        /* Wait for half a second */
+        std::this_thread::sleep_for(std::chrono::milliseconds(500));
```
why bother waiting?

my worry in having this here is that customers are going to copy/paste it because they'll think it's important

graebm commented on samples/mqtt/pkcs11_connect/main.cpp:
```diff
@@ -199,99 +179,24 @@ int main(int argc, char *argv[])
     }
 
-    // wait for the OnConnectionCompleted callback to fire, which sets connectionCompletedPromise...
-    if (connectionCompletedPromise.get_future().get() == false)
+    if (connectionCompletedPromise.get_future().get())
     {
-        fprintf(stderr, "Connection failed\n");
-        exit(-1);
-    }
```
[style] go back to the old way of early-exit, instead of embedding the rest of the main program in an indented block.

early-exit is much easier to read
other people have other opinions, but they're wrong

TwistedTwigleg commented:
Thanks for the review! I modified the code accordingly 👍

graebm commented on .builder/actions/build_samples.py:
```diff
@@ -11,9 +11,12 @@ def run(self, env):
+            'samples/mqtt/pkcs11_connect',
+            'samples/mqtt/raw_connect',
+            'samples/mqtt/websocket_connect',
+            'samples/mqtt/x509_connect',
```
we might want a different name for the "x509_connect", since the "basic_connect" is also technically using X.509 certificates.

The difference is that "basic_connect" sample uses X.509 certificates during the TLS handshake.

While the "x509_connect" sample uses basic TLS (only the server's identity is authenticated) during the TLS handshake, but then an X.509 certificate is used to prove the client's identity during the websocket handshake. Maybe call this "websocket_x509_connect"? Or "x509_cred_provider_connect"?

Run it by Bret, he knows more about credentials providers

bretambrose replied:
It's wordy, but x509_credentials_provider_connect is accurate.

TwistedTwigleg replied:
I'll change it to "x509_credentials_provider_connect" 👍

bretambrose commented on codebuild/samples/pkcs11-pubsub-linux.sh:
```diff
@@ -24,14 +24,14 @@ openssl pkcs8 -topk8 -in /tmp/privatekey.pem -out /tmp/privatekey.p8.pem -nocryp
 
 # build and run sample
-pushd $CODEBUILD_SRC_DIR/samples/mqtt/pkcs11_pub_sub
+pushd $CODEBUILD_SRC_DIR/samples/mqtt/pkcs11_connect
```
Maybe these scripts should be renamed now that we're saying "connect" instead of "pubsub"

bretambrose commented on samples/README.md:
```diff
@@ -1,8 +1,9 @@
+* [Basic Pub-Sub](#basic-pub-sub)
+* [Websocket Connect](#websocket-connect)
+* [PKCS#11 Connect](#pkcs11-connect)
+* [Raw Connect](#raw-connect)
```
Can we add an entry for the X509 connect sample?

bretambrose commented on samples/utils/CommandLineUtils.cpp:
```diff
@@ -162,13 +174,266 @@ namespace Utils
+            Aws::Iot::MqttClientConnectionConfigBuilder(certificatePath.c_str(), keyPath.c_str());
+        clientConfigBuilder.WithEndpoint(endpoint);
+
+        if (HasCommand("ca_file"))
```
I would move all of these raw strings into constants so the string itself isn't repeated all over

TwistedTwigleg commented:
Thanks for the reviews! I adjusted the code accordingly 👍

bretambrose commented on samples/README.md:
```diff
@@ -164,27 +221,106 @@ To run this sample using [SoftHSM2](https://www.opendnssec.org/softhsm/) as the
+## Raw Connect
+
+This sample is similar to the [Basic Connect](#basic-connect), but the connection setup is more manual.
 This is a starting point for using custom
```
custom auth is another use case here that could be linked

bretambrose commented on samples/README.md:
```diff
@@ -164,27 +221,106 @@ To run this sample using [SoftHSM2](https://www.opendnssec.org/softhsm/) as the
+## x509 Credentials Provider Connect
+
+This sample is similar to the [Basic Pub-Sub](#basic-pub-sub), but the connection uses a X.509 certificate
+to prove client identity during the websocket handshake rather than during the TLS handshake.
```
This isn't quite accurate.  The X509 certificate is used to source AWS credentials.  The websocket handshake is not change in any way, just the credential source.

bretambrose commented on samples/mqtt/x509_credentials_provider_connect/main.cpp:
```diff
@@ -0,0 +1,247 @@
+
+    builder.WithEndpoint(endpoint);
+
+    auto clientConfig = builder.Build();
```
For the connect samples, isn't everything the same after the client config construction?  Can we refactor that to not be repeated?

Along the same lines, why don't these samples use the utils connection builder helper functions?

TwistedTwigleg commented:
Thanks for the review! I modified the code accordingly. Now the connection samples use the utility functions and I moved the code for connecting and disconnecting to CommandLineUtils - now the samples are super compact and tidy 🙂 

Will apply the same changes to the other PRs as well.

graebm commented:
You'll need to merge in and rewrite the new windows_cert_pub_sub sample

TwistedTwigleg commented:
I pulled `main` and modified the Windows Cert sample 👍

sbSteveK commented on samples/README.md:
````diff
@@ -228,26 +307,84 @@ To run this sample with a basic certificate from AWS IoT Core:
 
     ```
-    .\windows-cert-pub-sub.exe --endpoint xxxx-ats.iot.xxxx.amazonaws.com --ca_file AmazonRootCA.pem --cert CurrentUser\My\A11F8A9B5DF5B98BA3508FBCA575D09570E0D2C6
+    .\windows-cert-connect.exe --endpoint xxxx-ats.iot.xxxx.amazonaws.com --ca_file AmazonRootCA.pem --cert CurrentUser\My\A11F8A9B5DF5B98BA3508FBCA575D09570E0D2C6
````
Maybe these commands should look like the others for user input?

TwistedTwigleg commented:
Thanks for the review! Merging into main...
